### PR TITLE
Migrate variants with nil unit_value

### PR DESCRIPTION
**Pull request by andrewpbrett**
#### What? Why?

Addresses #6737

Variants with a `unit_value` of `nil` and whose product uses `items` cannot be changed to use `weight` as their `variant_unit`. We have some assumptions in the code that `unit_value` is never `nil`.

#### What should we test?

The issue described in #6737 will no longer be reproducible <s>without manually altering the database directly</s> even if you try to alter the database directly. 

#### Release notes

Changelog Category: User facing changes | Technical changes

### CODE REVIEW

File: db/migrate/20210202052337_migrate_variant_unit_values.rb
Comment (sauloperez) [inline db/migrate/20210202052337_migrate_variant_unit_values.rb:5]: This will literally blow up the server because we are loading heaps of records into memory. We should rather query the ones we care about by joining variants and products together.

File: db/migrate/20210202052337_migrate_variant_unit_values.rb
Comment (sauloperez) [inline db/migrate/20210202052337_migrate_variant_unit_values.rb:4]: This will literally blow up the server because we are loading heaps of records into memory. We should rather query the ones we care about by joining variants and products together.

File: db/migrate/20210202052337_migrate_variant_unit_values.rb
Comment (sauloperez) [inline db/migrate/20210202052337_migrate_variant_unit_values.rb:7]: I guess we need some callbacks to be executed, right? If not, running a single update statement using `#update_all` would be best.

**Review by sauloperez — CHANGES_REQUESTED** [review]

**Comment by sauloperez** [comment]
> We have some assumptions in the code that unit_value is never nil.

is there something preventing us from enforcing that at DB-level? without a DB constraint, we can't guarantee so.

File: db/migrate/20210202052337_migrate_variant_unit_values.rb
Comment (andrewpbrett) [reply db/migrate/20210202052337_migrate_variant_unit_values.rb:8]: I think we do want the callbacks; I at least erred on the side of running them here since there's quite a few of them and I don't feel like I fully understand what's happening with all of them.

File: db/migrate/20210202052337_migrate_variant_unit_values.rb
Comment (andrewpbrett) [reply db/migrate/20210202052337_migrate_variant_unit_values.rb]: whoops :) 👍

**Comment by andrewpbrett** [comment]
> is there something preventing us from enforcing that at DB-level?

I don't think so, I added a line to the migration that puts the constraint there.

File: db/migrate/20210202052337_migrate_variant_unit_values.rb
Comment (sauloperez) [inline db/migrate/20210202052337_migrate_variant_unit_values.rb:6]: Sorry I didn't mention it but we'll also need to use `#find_each` if can't avoid instantiating the variants to run callbacks.

File: db/migrate/20210202052337_migrate_variant_unit_values.rb
Comment (sauloperez) [inline db/migrate/20210202052337_migrate_variant_unit_values.rb:10]: 💯 this is gold!

Comment (andrewpbrett) [reply db/migrate/20210202052337_migrate_variant_unit_values.rb:10]: 👍 😁

**Review by sauloperez — APPROVED** [review]

**Comment by sauloperez** [comment]
Now I noticed we haven't updated the db/schema's timestamp. This needs the be executed locally and the schema updated. I move it to Dev test in case you want to play with it in a staging server @andrewpbrett

**Comment by andrewpbrett** [comment]
@sauloperez I tested it out; I created a Variant with a `nil` `unit_value` and confirmed that the migration set it to be `1`; further, it's no longer possible to set `unit_value` to be `nil` after this migration. 👍 Do we want a second review or good to merge now?

**Comment by sauloperez** [comment]
That's enough for me.

File: db/migrate/20210202052337_migrate_variant_unit_values.rb
Comment (sauloperez) [inline db/migrate/20210202052337_migrate_variant_unit_values.rb:14]: :heart_eyes: finally we leverage PostgreSQL features

**Review by sauloperez — APPROVED** [review]

File: db/migrate/20210202052337_migrate_variant_unit_values.rb
Comment (sauloperez) [inline db/migrate/20210202052337_migrate_variant_unit_values.rb:3]: lol I didn't know `NAN` existed

Comment (Matt-Yorkley) [reply db/migrate/20210202052337_migrate_variant_unit_values.rb:3]: I know it exists in Rails, but I'm surprised Postgres allows saving it as a value on a decimal field in the DB...

Comment (sauloperez) [reply db/migrate/20210202052337_migrate_variant_unit_values.rb:3]: you gotta love the :elephant: :heart:

**Pull request merged**
